### ISSUE TITLE
getSeq(key) API (or similar)

### ISSUE THREAD

**Issue by regular** [ssbc/scuttlebot#543]
## Would people accept a PR that exposes a way to query the internal [db.keys index](https://github.com/ssbc/secure-scuttlebutt/blob/master/db.js#L8) from a plugin?

I have a [use case](https://github.com/regular/ssb-revisions) that seems like needing it to work efficiently, and currently it is impossible to get the seq (as in flumelog-offset, not as in feed sequence) from a message key (%...).

Are there any plans for the interface between plugins and flumedb? Currently there's `_flumeUse()`, which is labeled as a temporary hack. It exposes the flumelog to plugins (via createView), but not the flumedb itself.

The concrete case where I ran into this issue is an attempt to index mutable messages. Something like messagesByType, but including only the latest version of each message. The idea is to have a drop-in replacement for flumeview-level, that will additionally _remove_ entries from the index that are _no longer valid_. For this to work, the flumeview-level replacement (I call it flume*re*view-level) has to feed the old and the new message value into the separate calls to the map function and then compute a diff of the resulting arrays. The old value might be a previous revision or the original message (revisionRoot). In order to tell if a given index already has indexed the original message or not, I need to compare the original messages' seq against the index's `since` value. However, for efficiency sake, I do not store seqs for original messages (for revisions I do), the idea being to not add a lot of overhead to ssb networks that have millions of messages but only thousands of revisions.

Another way to put it: When a revision comes in, the plugin learns, after the fact, that it should have remembered a particular message's seq number. Now all it has it that messages key.

Currently my only options seems to be to either duplicate the [keys index](https://github.com/ssbc/secure-scuttlebutt/blob/master/db.js#L8) in my plugin or to fork scuttlebot, both of which I'd prefer not to.

**Comment by cryptix** [comment]
Dominic showed me a plugin that uses more native features of flume. I think it was this one: https://github.com/ssbc/ssb-search

As someone that currently implements ssb on go, I’m a bit weary/partial on this. I see the useful-ness. I just fear back-porting such Features if they become widely adopted but I‘m all for isle experiments, so to speak...

**Comment by regular** [comment]
@cryptix I wouldn't worry about portability for this. After all, any implementation will have a mapping between msg keys and some monotonically growing offset/record number into the local append-only log, or are you looking into doing something fundamentally different?

Thanks! I checked out flumeview-search, but it is pretty standard stuff. It accesses the log like the other views do too.

**Comment by dominictarr** [comment]
@regular sure. probably we should just have it so a plugin can see the flume instance, and call various methods (which may not be exposed over muxrpc). I've also considered something like flumeview-reduce - but a key-value reduce, where there are lots of small views (say, each being a snapshot of a particular mutable state, say a thread or other document) and these would then be stored in a key-value store like level.

**Comment by regular** [comment]
@dominictarr Just realised that exposing flumedb would not be enough to implement getSeq(), because even if you have reference to the flumeview-hashtable instance, there's no way to get to the actual content of the hashtable (it's mapped through log.get), so this would require some more drilling of holes ...

Yes, I think it would make sense to expose flumdb to plugins while protecting it agains remote access.

I actually tried to implement the special flumeview-reduce you described (I called it flumeview-group-reduce), there would be a ton of use cases for it, I reckon, even stuff like about messages) but didn't get very far with it, then switched to implement ssb-revisions on top of a regular flumeview-reduce for simplicity's sake.

I wrote a little not to myself back then:

> What we eventually want, is a per-revRoot index of all revisions and, in particular, the heads. This is a reduce, but for a group of messages (the ones with the same revRoot)

> One way we could achieve this:

> Implement a flumeview-reduce that maintains its own leveldb with one record per revRoot. The record contains the list of revisions and the heads, and/or roots, other meta-data

> Since reduce needs to be a sync function, it will have to maintain a write-queue to write to leveldb asynchronously: It builds up batch writes until a previous write succeeds. (use async-single)

> This way, writes of the index should be much more efficient. (TODO: benchmark)

**Comment by dominictarr** [comment]
do you have code left over from that?

How I'd probably do it: have a cache of reduce objects you currently have in memory. when you see a message, get it's object, if it's in the cache it will be fast. when the number of "dirty" objects hits some limit, or nothing has changed in a while, write them in a batch.

**Comment by dominictarr** [comment]
I don't have anything against flumeview-hashtable exposing sequences, but I just havn't needed it before now.

**Comment by regular** [comment]
@dominictarr I looked for it but wasn't able to find any code for that group-index thing (wasn't that great anyway). I'd probably base it on flume-level now, with group members being neighbours  in leveldb key space. That's how I solved the problem that made me open this issue anyway, no need for an additional sbot API anymore. I reduce at query-time now, which is fast enough and will allow for subjective and dynamic edit permissions in the future. That code is [here](https://github.com/regular/ssb-revisions/blob/master/index.js )
